File: HolidayShowLib/ByteParserBase.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace HolidayShowLib
{
    public abstract class ByteParserBase 
    {
        /// <summary>
        ///  Used to syncronize the buffer list.
        /// </summary>
        private readonly object _byteListLock = new object();
        /// <summary>
        /// Buffer size to start the memory stream.  This is only to start, and the stream will add to it as it needs.
        /// </summary>
        private const int STARTING_BUFFER_SIZE = 2048;

        /// <summary>
        /// Stream to hold the data as its received from the medium.
        /// </summary>
        private readonly MemoryStream _messageBuffer = new MemoryStream(STARTING_BUFFER_SIZE);

        /// <summary>
        /// Contains a list of possible parsing protocols - Handy if more than one protocols comes across the line.
        /// </summary>
        protected readonly List<ParserProtocolContainer> Parsers = new List<ParserProtocolContainer>();

        private readonly List<BytePositions> ParserResults = new List<BytePositions>();

        private bool _isDisposed;


        public void BytesReceived(byte[] byteBuffer)
        {
            if (ParserResults.Count != Parsers.Count)
            {
                //Clear for safety
                ParserResults.Clear();

                // Initilize the parsers
                foreach (var parser in Parsers)
                {
                    ParserResults.Add(new BytePositions(parser));
                }
            }

            // It may be considered that here we should not block, and return to the caller asap, and add data to the list async, but in order
            lock (_byteListLock)
            {
                // Write the data to the memory stream
                _messageBuffer.Write(byteBuffer, 0, byteBuffer.Length);
#if DEBUG
                buffersReceived++;
#endif
            }

            // Initiate a buffer search
            SearchBuffer();
        }

        [DebuggerDisplay("Start={Start};End={End};EndOnly={EndParserOnly}")]
        private class BytePositions
        {
            public BytePositions(ParserProtocolContainer parser)
            {
                this.Parser = parser;

                EndParserOnly = parser.StartingBytes.Length == 0 && parser.EndingBytes.Length > 0;
                Start = EndParserOnly ? 0 : -1;  // If its an end-only, we want the starting position to be the first element of the stream all the time.
                End = -1;
            }

            public ParserProtocolContainer Parser { get; private set; }
            public int Start { get; set; }
            public int End { get; set; }

            public bool EndParserOnly { get; private set; }

            public void Reset()
            {
                Start = EndParserOnly ? 0 : -1;  // If its an end-only, we want the starting position to be the first element of the stream all the time.
                End = -1;
            }
        }

#if DEBUG

        private long buffersReceived;
        private long bufferSearches;
        private long buffersReturned;

#endif

        private void SearchBuffer()
        {
  
            int searchAgain = 0;
            var stopWatch = Stopwatch.StartNew();

        top:

#if DEBUG
            bufferSearches++;
#endif

            bool searchBufferAgain = false;
            lock (_byteListLock)
            {
                try
                {

                    for (var i = 0; i < ParserResults.Count; i++)
                    {
                        // Get the parser results, and reset the values between each pass
                        var p = ParserResults[i];
                        var parser = p.Parser;
                        p.Reset();

                        // There may be no initial byte[], and we are only splitting on a sequence of end bytes.  
                        // This feature is new in DataEvents 5.14.0501.0 and greater.  
                        if (!p.EndParserOnly)
                        {
                            p.Start = FindPosition(parser.StartingBytes);

                            if (p.Start == -1)
                            {
                                continue;
                            }
                        }

                        // Offset the end find by the last start found, and the starting bytes length
                        var startingPositition = p.Start + parser.StartingBytes.Length;

                        // Since a start was found, find the next position, after the start with the ending sequence
                        p.End = FindPosition(startingPositition, parser.EndingBytes);
                    }

                    // First look for a full packet, at the lowest index start index found
                    var lowestParser = ParserResults.Where(x => x.End != -1).OrderBy(x => x.Start).FirstOrDefault();

                    // If no full packet is found, find the lowest parser where the start of the packet was found
                    if (lowestParser == null)
                    {
                        lowestParser = ParserResults.Where(x => x.Start > -1).OrderBy(x => x.Start).FirstOrDefault();
                    }

                    // If the lowest parser is still null, lets see if there is a EndOnly parser
                    if (lowestParser == null)
                    {
                        lowestParser = ParserResults.Where(x => x.EndParserOnly && x.End > 0).OrderBy(x => x.Start).FirstOrDefault();
                    }

                    if (
                        lowestParser == null ||
                        (!lowestParser.EndParserOnly && lowestParser.Start != 0 && lowestParser.End == -1)  // Look for any start that is found, without and end... and is not an EndOnly Parser
                        )
                    {
                        var newLength = 0;
                        var newStart = 0;

                        if (lowestParser != null && lowestParser.Start != -1)
                        {
                            newStart = lowestParser.Start;
                            newLength = (int)_messageBuffer.Length - newStart;
                        }

                        // Move the data left
#if WINDOWS_UWP
                        ArraySegment<byte> data;
                        if (_messageBuffer.TryGetBuffer(out data))
                        {
                            Buffer.BlockCopy(data.Array, newStart, data.Array, 0, newLength);
                        }
                        
#else
                        Buffer.BlockCopy(_messageBuffer.GetBuffer(), newStart, _messageBuffer.GetBuffer(), 0, newLength);
#endif

                        _messageBuffer.SetLength(newLength);
                        _messageBuffer.Position = newLength;

                        // nothing left to do, exit function
#if DEBUG && !WINDOWS_UWP
                        // goto end for debug parse output.
                        goto end;
#else
                        return;
#endif
                    }

                    if (lowestParser.Start != -1 && lowestParser.End != -1)
                    {
                        // The lowest index has a start / end buffer that 
                        var parser = lowestParser.Parser;

                        // The final byte index with the sequence value added in.
                        var realEnd = lowestParser.End + parser.EndingBytes.Length;

                        // Sets the length of the expected data - the entire packet
                        var messagelength = realEnd - lowestParser.Start;

                        // Selects the data that we are expecting - starting bytes to end of ending sequence bytes
                        //var bytesRead = _messageBuffer.Select(x => x).Skip(byteStart).Take(messagelength).ToArray();
                        var bytesRead = new byte[messagelength];
#if WINDOWS_UWP

                        ArraySegment<byte> bytes;
                        if (_messageBuffer.TryGetBuffer(out bytes))
                        {
                            Buffer.BlockCopy(bytes.Array, lowestParser.Start, bytesRead, 0, messagelength);
                        }

#else

                        Buffer.BlockCopy(_messageBuffer.GetBuffer(), lowestParser.Start, bytesRead, 0, messagelength);
#endif

                        if (bytesRead != null)
                        {
                            // Sends off for processing
                            ProcessPacket(bytesRead, parser);
#if DEBUG
                            buffersReturned++;
#endif
                        }

                        // truncates the messageBuffer
                        var length = (int)_messageBuffer.Length - realEnd;
                        if (length != 0)
                        {
#if WINDOWS_UWP
                            ArraySegment<byte> data;
                            if (_messageBuffer.TryGetBuffer(out data))
                            {
                                Buffer.BlockCopy(data.Array, realEnd, data.Array, 0, length);
                            }
                            // Reset the stream to the new end
                            
#else
                            Buffer.BlockCopy(_messageBuffer.GetBuffer(), realEnd, _messageBuffer.GetBuffer(), 0, length);
#endif
                        }

                        _messageBuffer.SetLength(length);
                        _messageBuffer.Position = length;

                        // If there is more bytes to look for, search again (2 would mean a min of two bytes + a data byte at a min)
                        if (length > 3)
                            searchBufferAgain = true;
                    }
                }
                catch
                {
                    _messageBuffer.Position = 0;
                }
            }

            // outside the lock to prevent a deadlock.
            if (searchBufferAgain)
            {
                searchAgain++;
                //_loggerService.Debug("SearchAgain!");
                goto top; // do this until the entire buffer is processed
            }


#if DEBUG && !WINDOWS_UWP
            end:
            Console.WriteLine("{0}; ReSearch {1}; Received: {2}; Searches: {3}; Returned: {4}", stopWatch.Elapsed, searchAgain, buffersReceived, bufferSearches, buffersReturned);
#endif
        }

        public abstract void ProcessPacket(byte[] bytes, ParserProtocolContainer parser);

        private static bool NextBytesMatch(int pos, byte[] searchBytes, Byte[] byteSequence)
        {
            for (int i = 0; i < byteSequence.Length; i++)
            {
                if (searchBytes[pos + i] != byteSequence[i])
                {
                    return false;
                }
            }

            return true;
        }


        private int FindPosition(int startingPos, byte[] byteSequence)
        {
            if ((byteSequence.Length + startingPos) > _messageBuffer.Length)
                return -1;

#if WINDOWS_UWP
            ArraySegment<byte> data;
            if (!_messageBuffer.TryGetBuffer(out data))
            {
                return -1;
            }
            var memoryBuffer = data.Array;
#else
            var memoryBuffer = _messageBuffer.GetBuffer();
#endif

            var membufferlen = _messageBuffer.Length;

            for (int i = startingPos; i < membufferlen; i++)
            {
                if (NextBytesMatch(i, memoryBuffer, byteSequence))
                    return i;
            }

            return -1;
        }


        private int FindPosition(byte[] byteSequence)
        {
            return FindPosition(0, byteSequence);
        }

    }
}

File: HolidayShowLibNative/ByteParserBase.cpp

#ifdef _WIN32
#include "stdafx.h"
#endif

#include <cstring>
#include <iostream>
#include <stdio.h>
#include "ByteParserBase.h"
#include <algorithm>

using namespace std;

namespace HolidayShowLib
{
	
	

	bool ByteParserBase::NextBytesMatch(int pos, ByteBuffer& searchBytes, ByteBuffer& byteSequence)
	{
		for (int i = 0; i < byteSequence.size(); i++)
		{
			if (searchBytes[pos + i] != byteSequence[i])
			{
				return false;
			}
		}

		return true;
	}

	int ByteParserBase::FindPosition(int startingPos, ByteBuffer& byteSequence)
	{

		if ((byteSequence.size() + startingPos) > _messageBuffer.size())
			return -1;

		for (int i = startingPos; i < _messageBuffer.size(); i++)
		{
			if (NextBytesMatch(i, _messageBuffer, byteSequence))
				return i;
		}

		return -1;

	}

	ByteParserBase::ByteParserBase()
	{
	}


	ByteParserBase::~ByteParserBase()
	{
	}

	void ByteParserBase::BytesReceived(ByteBuffer& byteBuffer)
	{
		if (_parser.size() != _parserResults.size())
		{
			//Clear for safety
			_parserResults.clear();



			// Initilize the parsers
			for (auto p = 0; p < _parser.size(); p++)
			{
				auto parser = _parser[p];
				_parserResults.emplace_back(std::make_shared<BytePositions>(parser));
			}
		}

		// It may be considered that here we should not block, and return to the caller asap, and add data to the list async, but in order
		{
			std::lock_guard<std::mutex> lock(_byteListLock);

			_messageBuffer.insert(std::end(_messageBuffer), std::begin(byteBuffer), std::end(byteBuffer));
		}

		// Initiate a buffer search
		SearchBuffer();
	}

	void ByteParserBase::SearchBuffer()
	{

		bool searchBufferAgain = false;
		
		{
			std::lock_guard<std::mutex> lock(_byteListLock);

			try
			{
				for (auto i = 0; i < _parserResults.size(); i++)
				{
					// Get the parser results, and reset the values between each pass
					auto p = _parserResults[i];
					auto parser = p->ParserGet();
					p->Reset();

					// There may be no initial byte[], and we are only splitting on a sequence of end bytes.  
					// This feature is new in DataEvents 5.14.0501.0 and greater.  
					if (!p->EndParserOnlyGet())
					{
						p->StartSet(FindPosition(0, parser->StartingBytesGet()));

						if (p->StartGet() == -1)
						{
							continue;
						}
					}

					// Offset the end find by the last start found, and the starting bytes length
					auto startingPositition = p->StartGet() + parser->StartingBytesGet().size();

					// Since a start was found, find the next position, after the start with the ending sequence
					p->EndSet(FindPosition(startingPositition, parser->EndingBytesGet()));
				}

                // Presort the list vector 
				std::sort(std::begin(_parserResults), std::end(_parserResults), [](const BytePositionPtr& a, const BytePositionPtr& b)
				{
					return a->StartGet() < b->StartGet();
				});

             
				auto lowestParser = std::find_if(std::begin(_parserResults), std::end(_parserResults), [](const BytePositionPtr& p)
				{
					return p->EndGet() != -1;
				});  

                if (lowestParser == _parserResults.end())
                {
					lowestParser = std::find_if(std::begin(_parserResults), std::end(_parserResults), [](const BytePositionPtr& p)
					{
						return p->StartGet() > -1;
					});
				}

				if (lowestParser == _parserResults.end())
				{
					lowestParser = std::find_if(std::begin(_parserResults), std::end(_parserResults), [](const BytePositionPtr& p)
					{
						return p->EndParserOnlyGet() && p->EndGet() > 0;
					});
				}


				if (lowestParser  == _parserResults.end() ||
					(!lowestParser[0]->EndParserOnlyGet() && lowestParser[0]->StartGet() != 0 && lowestParser[0]->EndGet() == -1)  // Look for any start that is found, without and end... and is not an EndOnly Parser
					)
				{
					auto newLength = 0;
					auto newStart = 0;

					if (lowestParser != std::end(_parserResults) && lowestParser[0]->StartGet() != -1)
					{
						newStart = lowestParser[0]->StartGet();
						newLength = (int)_messageBuffer.size() - newStart;
					}

					memmove(_messageBuffer.data(), _messageBuffer.data() + newStart, newLength);
					// Move the data left
					//Buffer.BlockCopy(_messageBuffer.GetBuffer(), newStart, _messageBuffer.GetBuffer(), 0, newLength);
					_messageBuffer.resize(newLength);

					return;
				}

				if (lowestParser[0]->StartGet() != -1 && lowestParser[0]->EndGet() != -1)
				{
					// The lowest index has a start / end buffer that 
					auto parser = lowestParser[0]->ParserGet();

					// The final byte index with the sequence value added in.
					auto realEnd = lowestParser[0]->EndGet() + parser->EndingBytesGet().size();

                    // starting byte location
                    auto realStart = lowestParser[0]->StartGet();

					// Sets the length of the expected data - the entire packet
					auto messagelength = realEnd - realStart;

					// Selects the data that we are expecting - starting bytes to end of ending sequence bytes
					//var bytesRead = _messageBuffer.Select(x => x).Skip(byteStart).Take(messagelength).ToArray();
					//auto bytesRead = new byte[messagelength];
					ByteBuffer bytesRead;

					bytesRead.insert(std::begin(bytesRead), std::begin(_messageBuffer) + realStart, std::begin(_messageBuffer) + realStart + messagelength);

					//Buffer.BlockCopy(_messageBuffer.GetBuffer(), lowestParser.Start, bytesRead, 0, messagelength);

					// Sends off for processing
					ProcessPacket(bytesRead, parser);

					// truncates the messageBuffer
					auto length = (int)_messageBuffer.size() - realEnd;
					if (length != 0)
					{
						memmove(_messageBuffer.data(), _messageBuffer.data() + realEnd, length);

						// Reset the stream to the new end
						//Buffer.BlockCopy(_messageBuffer.GetBuffer(), realEnd, _messageBuffer.GetBuffer(), 0, length);
					}

					_messageBuffer.resize(length);

					// If there is more bytes to look for, search again (2 would mean a min of two bytes + a data byte at a min)
					if (length > 3)
						searchBufferAgain = true;
				}
			}
			catch (const std::exception& ex)
			{
				
			}
			
		}


	}


};